File: app/src/main/java/com/labessence/biotech/finaldilution/peripherals/datastores/SharedPreferencesStore.java
package com.labessence.biotech.finaldilution.peripherals.datastores;

import android.content.SharedPreferences;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import com.labessence.biotech.finaldilution.genericitem.Item;
import com.labessence.biotech.finaldilution.peripherals.DataGatewayOperations;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * Project: FinalDilution
 * Created by dawid.chmielewski on 4/3/2017.
 */

public class SharedPreferencesStore<T extends Item> implements DataGatewayOperations<T> {

    private static final String TAG = "File DataGateway";
    //private static final File COMPOUNDS_FILE = new File("compounds.json");
    private final SharedPreferences sharedPreferences;
    private Gson gson = new GsonBuilder().create();
    private TypeToken<List<T>> listType;


    //TODO gson.toJsonTree(compound);
    // if found - load, if not create.
    // not found : no solutions.
    // solution created - saved.


    public SharedPreferencesStore(SharedPreferences sharedPreferences, TypeToken<List<T>> listType) {
        this.sharedPreferences = sharedPreferences;
        this.listType = listType;
    }

    @Override
    public void save(T item) {
        //TODO performance
        // Could use append
        // http://stackoverflow.com/questions/1625234/how-to-append-text-to-an-existing-file-in-java

        List<T> itemList = getItemListFormPreferences();
        if (itemList.contains(item)) {
            Log.d(TAG, "Item " + item.getName() + " already present.");
            return;
        }
        itemList.add(item);
        putItemListToPreferences(itemList);
    }


    @Override
    public void update(T item) {
        List<T> itemList = getItemListFormPreferences();
        //TODO Add means to update
        if (itemList.isEmpty()) {
            return;
        }

        Iterator<T> i = itemList.iterator();
        while (i.hasNext()) {
            T currentItem = i.next();
            if (currentItem.getName().equals(item.getName())) {
                i.remove();
            }
        }

        itemList.add(item);
        putItemListToPreferences(itemList);
    }

    @Override
    public void remove(T item) {
        List<T> itemList = getItemListFormPreferences();
        if (itemList.isEmpty()) {
            return;
        }
        itemList.remove(item);
        putItemListToPreferences(itemList);
    }

    @Override
    public T load(String name) {
        List<T> itemList = getItemListFormPreferences();
        if (itemList.isEmpty()) {
            return null;
        }
        for (T item : itemList) {
            if (item.getName().equals(name)) {
                return item;
            }
        }
        //TODO Add exception at the interface to be thrown in such cases
        return null;
    }

    @Override
    public List<T> loadAll() {
        return getItemListFormPreferences();
    }

    private List<T> getItemListFormPreferences() {

        String rootJson = sharedPreferences.getString("root", "");
        List<T> loadedList = gson.fromJson(rootJson, listType.getType());

        if (loadedList == null) {
            Log.d(TAG, "Loading from preferences failed.");
            return new ArrayList<>();
        }
        return loadedList;
    }

    private void putItemListToPreferences(List<T> itemList) {
        SharedPreferences.Editor prefsEditor = sharedPreferences.edit();
        String rootJson = gson.toJson(itemList);
        prefsEditor.putString("root", rootJson);
        prefsEditor.apply();
    }

}

File: app/src/main/java/com/labessence/biotech/finaldilution/peripherals/datastores/SharedPreferencesStore.kt
package com.labessence.biotech.finaldilution.peripherals.datastores

import android.content.SharedPreferences
import android.util.Log
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.labessence.biotech.finaldilution.component.concentration.Concentration
import com.labessence.biotech.finaldilution.genericitem.Item
import com.labessence.biotech.finaldilution.peripherals.DataGatewayOperations
import java.util.*

/**
 * Project: FinalDilution
 * Created by dawid.chmielewski on 4/3/2017.
 */

//TODO gson.toJsonTree(compound);
// if found - load, if not create.
// not found : no solutions.
// solution created - saved.

//private static final File COMPOUNDS_FILE = new File("compounds.json");
class SharedPreferencesStore<T : Item>(
    private val sharedPreferences: SharedPreferences, private val listType: TypeToken<List<T>>
) : DataGatewayOperations<T> {

    private val gson = GsonBuilder()
        .registerTypeAdapter(Concentration::class.java, ConcentrationDeserializer())
        .create()

    private val itemListFormPreferences: MutableList<T>
        get() {

            val rootJson = sharedPreferences.getString("root", "")
            val loadedList = gson.fromJson<List<T>>(rootJson, listType.type)

            if (loadedList == null) {
                Log.d(TAG, "Loading from preferences failed.")
                return ArrayList()
            }
            return loadedList as MutableList<T>
        }

    override fun size() = itemListFormPreferences.size


    override fun save(item: T) {
        //TODO performance
        // Could use append
        // http://stackoverflow.com/questions/1625234/how-to-append-text-to-an-existing-file-in-java

        val itemList = itemListFormPreferences
        if (itemList.contains(item)) {
            Log.d(TAG, "Item " + item.name + " already present.")
            return
        }
        itemList.add(item)
        putItemListToPreferences(itemList)
        Log.d(TAG, "Item " + item.name + " added.")
    }


    override fun rename(item: T, oldName: String) {
        val itemList = itemListFormPreferences
        if (itemList.isEmpty()) {
            return
        }

        removeItemByName(itemList, oldName)

        itemList.add(item)
        putItemListToPreferences(itemList)
        Log.d(TAG, "Item " + oldName + " renamed to " + item.name)
    }

    override fun update(item: T) {
        val itemList = itemListFormPreferences
        //TODO Add means to update
        if (itemList.isEmpty()) {
            return
        }

        removeItemByName(itemList, item.name)

        itemList.add(item)
        putItemListToPreferences(itemList)
        Log.d(TAG, "Item " + item.name + " updated.")
    }

    private fun removeItemByName(itemList: MutableList<T>, name: String) {
        val i = itemList.iterator()
        while (i.hasNext()) {
            val currentItem = i.next()
            if (currentItem.name == name) {
                i.remove()
            }
        }
    }

    override fun remove(item: T) {
        val itemList = itemListFormPreferences
        if (itemList.isEmpty()) {
            return
        }
        removeItemByName(itemList, item.name)
        putItemListToPreferences(itemList)
        Log.d(TAG, "Item " + item.name + " removed.")
    }

    override fun load(name: String): T? {
        val itemList = itemListFormPreferences
        if (itemList.isEmpty()) {
            return null
        }
        for (item in itemList) {
            if (item.name == name) {
                return item
            }
        }
        //TODO Add exception at the interface to be thrown in such cases
        return null
    }

    override fun loadAll(): List<T> {
        return itemListFormPreferences
    }

    private fun putItemListToPreferences(itemList: List<T>) {
        val prefsEditor = sharedPreferences.edit()
        val rootJson = gson.toJson(itemList)
        prefsEditor.putString("root", rootJson)
        prefsEditor.apply()
    }

    companion object {

        private val TAG = "File DataGateway"
    }

}
